xde/src/xde.rs

```diff
@@ -1535,6 +1551,60 @@ unsafe extern "C" fn xde_mc_tx(
     ptr::null_mut()
 }
 
+/// This is a generic wrapper for references that should be dropped once not in
+/// use.
+struct DropRef<DropFn, Arg>
+where
+    DropFn: Fn(*mut Arg) -> (),
+{
+    /// A function to drop the reference.
+    func: DropFn,
+    /// The reference pointer.
+    arg: *mut Arg,
+}
+
+impl<DropFn, Arg> DropRef<DropFn, Arg>
+where
+    DropFn: Fn(*mut Arg) -> (),
+{
+    /// Create a new `DropRef` for the provided reference argument. When this
+    /// object is dropped, the provided `func` will be called.
+    fn new(func: DropFn, arg: *mut Arg) -> Self {
+        Self { func, arg }
+    }
+
+    /// Return a pointer to the underlying reference.
+    fn inner(&self) -> *mut Arg {
+        self.arg
+    }
+}
+
+impl<DropFn, Arg> Drop for DropRef<DropFn, Arg>
+where
+    DropFn: Fn(*mut Arg) -> (),
+{
+    /// Call the cleanup function on the reference argument when we are dropped.
+    fn drop(&mut self) {
+        if !self.arg.is_null() {
+            (self.func)(self.arg);
+        }
+    }
+}
+
+// The following are wrappers for reference drop functions used in XDE.
+
+fn ire_refrele(ire: *mut ip::ire_t) {
+    unsafe { ip::ire_refrele(ire) }
+}
+
+fn nce_refrele(ire: *mut ip::nce_t) {
+    unsafe { ip::nce_refrele(ire) }
+}
+
+fn netstack_rele(ns: *mut ip::netstack_t) {
+    unsafe { ip::netstack_rele(ns) }
+}
```

Comment:
Nice, that's a useful abstraction. Perhaps it should live in a different mod to reuse it in other spots. But that can be a follow up.